database/writer.go

```diff
@@ -0,0 +1,167 @@
+
+	// MgoBulkWriter is a pipeline worker which properly batches bulk updates for MongoDB
+	MgoBulkWriter struct {
+		db           *DB              // provides access to MongoDB
+		conf         *config.Config   // contains details needed to access MongoDB
+		log          *log.Logger      // main logger for RITA
+		writeChannel chan BulkChanges // holds analyzed data
+		writeWg      *sync.WaitGroup  // wait for writing to finish
+		writerName   string           // used in error reporting
+		unordered    bool             // if the operations can be applied in any order, MongoDB can run the updates in parallel
+		maxBulkCount int              // max number of changes to include in each bulk update
+		maxBulkSize  int              // max total size of BSON documents making up each bulk update
+	}
+)
+
+// Size serializes the changes to BSON using provided buffer and returns total size
+// of the BSON description of the changes. Note this method slightly underestimates the
+// total amount BSON needed to describe the changes since extra flags may be sent along.
+func (m BulkChange) Size(buffer []byte) ([]byte, int) {
+	size := 0
+	if len(buffer) > 0 { // in case the byte slice has something in it already
+		buffer = buffer[:0]
+	}
+
+	if m.Selector != nil {
+		buffer, _ = bson.MarshalBuffer(m.Selector, buffer)
+		size += len(buffer)
+		buffer = buffer[:0]
+	}
+	if m.Update != nil {
+		buffer, _ = bson.MarshalBuffer(m.Update, buffer)
+		size += len(buffer)
+		buffer = buffer[:0]
+	}
+	return buffer, size
+}
+
+// Apply adds the change described to a bulk buffer
+func (m BulkChange) Apply(bulk *mgo.Bulk) {
+	if m.Selector == nil {
+		return // can't describe a change without a selector
+	}
+
+	if m.Remove && m.SelectAll {
+		bulk.RemoveAll(m.Selector)
+	} else if m.Remove /*&& !m.SelectAll*/ {
+		bulk.Remove(m.Selector)
+	} else if m.Update != nil && m.Upsert {
+		bulk.Upsert(m.Selector, m.Update)
+	} else if m.Update != nil && m.SelectAll {
+		bulk.UpdateAll(m.Selector, m.Update)
+	} else if m.Update != nil /*&& !m.Upsert && !m.SelectAll*/ {
+		bulk.Update(m.Selector, m.Update)
+	}
+}
+
+// NewBulkWriter creates a new writer object to write output data to collections
+func NewBulkWriter(db *DB, conf *config.Config, log *log.Logger, unorderedWritesOK bool, writerName string) *MgoBulkWriter {
+	return &MgoBulkWriter{
+		db:           db,
+		conf:         conf,
+		log:          log,
+		writeChannel: make(chan BulkChanges),
+		writeWg:      new(sync.WaitGroup),
+		writerName:   writerName,
+		unordered:    unorderedWritesOK,
+		maxBulkCount: 500,
+		maxBulkSize:  15 * 1000 * 1000,
+	}
+}
+
+// Collect sends a group of results to the writer for writing out to the database
+func (w *MgoBulkWriter) Collect(data BulkChanges) {
+	w.writeChannel <- data
+}
+
+// close waits for the write threads to finish
+func (w *MgoBulkWriter) Close() {
+	close(w.writeChannel)
+	w.writeWg.Wait()
+}
+
+// start kicks off a new write thread
+func (w *MgoBulkWriter) Start() {
```

Comment:
Please add comments throughout this function for clarity & readability

Reply:
Added